File: frontend/scripts/generate-favicon.py
#!/usr/bin/env python3
"""
Generate favicon files from SVG source.
Creates .ico file and various PNG sizes for different platforms.
"""

import subprocess
import os
from pathlib import Path

# Directory paths
script_dir = Path(__file__).parent
public_dir = script_dir.parent / "public"
svg_file = public_dir / "favicon.svg"
output_dir = public_dir

# Ensure output directory exists
output_dir.mkdir(parents=True, exist_ok=True)

# Define favicon sizes
# ICO format supports multiple sizes embedded
ico_sizes = [16, 32, 48]
png_sizes = [16, 32, 48, 64, 128, 180, 192, 512]

def check_dependencies():
    """Check if required tools are installed."""
    try:
        subprocess.run(["convert", "-version"], capture_output=True, check=True)
        return True
    except (subprocess.CalledProcessError, FileNotFoundError):
        print("ImageMagick is not installed or 'convert' command is not available.")
        print("Please install ImageMagick:")
        print("  Ubuntu/Debian: sudo apt-get install imagemagick")
        print("  macOS: brew install imagemagick")
        print("  Or use an online converter to convert the SVG to ICO manually.")
        return False

def generate_png(size):
    """Generate PNG file of specific size from SVG."""
    output_file = output_dir / f"favicon-{size}x{size}.png"
    cmd = [
        "convert",
        "-background", "none",
        "-density", "300",
        str(svg_file),
        "-resize", f"{size}x{size}",
        str(output_file)
    ]
    try:
        subprocess.run(cmd, check=True, capture_output=True)
        print(f"✓ Generated {output_file.name}")
        return output_file
    except subprocess.CalledProcessError as e:
        print(f"✗ Failed to generate {output_file.name}: {e.stderr.decode()}")
        return None

def generate_ico():
    """Generate ICO file with multiple sizes embedded."""
    # First generate PNGs for ICO
    temp_pngs = []
    for size in ico_sizes:
        png_file = generate_png(size)
        if png_file:
            temp_pngs.append(png_file)

    if not temp_pngs:
        print("✗ Failed to generate ICO: no PNGs created")
        return

    # Combine PNGs into ICO
    ico_file = output_dir / "favicon.ico"
    cmd = ["convert"] + [str(p) for p in temp_pngs] + [str(ico_file)]

    try:
        subprocess.run(cmd, check=True, capture_output=True)
        print(f"✓ Generated {ico_file.name}")
    except subprocess.CalledProcessError as e:
        print(f"✗ Failed to generate ICO: {e.stderr.decode()}")

def generate_apple_touch_icon():
    """Generate apple-touch-icon.png (180x180)."""
    output_file = output_dir / "apple-touch-icon.png"
    cmd = [
        "convert",
        "-background", "none",
        "-density", "300",
        str(svg_file),
        "-resize", "180x180",
        str(output_file)
    ]
    try:
        subprocess.run(cmd, check=True, capture_output=True)
        print(f"✓ Generated {output_file.name}")
    except subprocess.CalledProcessError as e:
        print(f"✗ Failed to generate apple-touch-icon: {e.stderr.decode()}")

def main():
    print("🎨 Generating favicon files from SVG...")
    print(f"Source: {svg_file}")
    print()

    if not svg_file.exists():
        print(f"✗ SVG file not found: {svg_file}")
        return

    if not check_dependencies():
        print("\n⚠️  Cannot generate favicons automatically.")
        print(f"The SVG file is available at: {svg_file}")
        print("You can use an online converter like:")
        print("  - https://convertio.co/svg-ico/")
        print("  - https://cloudconvert.com/svg-to-ico")
        return

    print("Generating ICO file...")
    generate_ico()
    print()

    print("Generating PNG files...")
    for size in png_sizes:
        if size not in ico_sizes:  # Skip sizes already generated
            generate_png(size)
    print()

    print("Generating Apple Touch Icon...")
    generate_apple_touch_icon()
    print()

    print("✅ Favicon generation complete!")
    print(f"Files saved to: {output_dir}")

if __name__ == "__main__":
    main()

File: frontend/scripts/generate-favicon.js
#!/usr/bin/env node
/**
 * Generate favicon files from SVG source.
 * Creates .ico-compatible PNG and various PNG sizes for different platforms.
 */

import sharp from 'sharp';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const publicDir = path.join(__dirname, '..', 'public');
const svgFile = path.join(publicDir, 'favicon.svg');

// Define sizes to generate
const sizes = [
  { name: 'favicon-16x16.png', size: 16 },
  { name: 'favicon-32x32.png', size: 32 },
  { name: 'favicon-48x48.png', size: 48 },
  { name: 'favicon.ico', size: 32 }, // Standard favicon.ico size
  { name: 'apple-touch-icon.png', size: 180 },
  { name: 'android-chrome-192x192.png', size: 192 },
  { name: 'android-chrome-512x512.png', size: 512 },
];

async function generateFavicon(svgBuffer, name, size) {
  const outputPath = path.join(publicDir, name);

  try {
    await sharp(svgBuffer, { density: 300 })
      .resize(size, size, {
        fit: 'contain',
        background: { r: 0, g: 0, b: 0, alpha: 0 }
      })
      .png()
      .toFile(outputPath);

    console.log(`✓ Generated ${name} (${size}x${size})`);
    return true;
  } catch (error) {
    console.error(`✗ Failed to generate ${name}:`, error.message);
    return false;
  }
}

async function generateFavicons() {
  console.log('🎨 Generating favicon files from SVG...');
  console.log(`Source: ${svgFile}\n`);

  if (!fs.existsSync(svgFile)) {
    console.error(`✗ SVG file not found: ${svgFile}`);
    process.exit(1);
  }

  const svgBuffer = fs.readFileSync(svgFile);

  let successCount = 0;
  let failCount = 0;

  // Generate each favicon sequentially
  for (const { name, size } of sizes) {
    const success = await generateFavicon(svgBuffer, name, size);
    if (success) {
      successCount++;
    } else {
      failCount++;
    }
  }

  console.log('\n' + '='.repeat(50));
  console.log(`✅ Favicon generation complete!`);
  console.log(`   Success: ${successCount}/${sizes.length}`);
  if (failCount > 0) {
    console.log(`   Failed: ${failCount}`);
  }
  console.log(`   Files saved to: ${publicDir}`);
  console.log('='.repeat(50));
}

generateFavicons().catch(error => {
  console.error('Error generating favicons:', error);
  process.exit(1);
});
